addons/mail/models/mail_thread.py

```diff
@@ -1005,6 +1006,63 @@ def message_route(self, message, message_dict, model=None, thread_id=None, custo
             (email_from, email_to, message_id)
         )
 
+    def _ignore_incoming_email(self, message, message_dict, routes):
+        """This method returns True is the incoming email should be ignored.
+
+        The goal of this method is to prevent loops which can occur if an auto-replier
+        send email to Odoo.
+
+        It also detects common headers set by the auto-repliers / mailing lists.
+        """
+
+        # Detect if the email has been sent by an auto-replier or by a mailing list
+        for header in ('X-Autoreply', 'X-Autorespond', 'Feedback-ID', 'List-Id', 'List-Unsubscribe'):
```

Comment:
I feel like we could split between two main content: a generic one (based on headers / message), then a model-based one (using the routes). That way we could implement some specific behavior. Basic code in model-based method would be simply to do a search_count . Like
```
@api.model
def _routing_detect_loop_model(self, message, message_dict, routes):
    # do generic stuff
    ....

    for model, thread_id, custom_values, user_id, alias in routes or []:
        self.env(model)._routing_detect_loop(thread_id, custom_values, user_id, alias)

def _routing_detect_loop(self, thread_id, custom_values, user_id, alias):
    if thread_id:
        return False
    mail_incoming_messages_count = model.sudo().search_count([])
    if mail_incoming_messages_count >= INCOMING_LIMIT_ALIAS:
        _logger.info('This email address created too many <%s>.', description)
        return True
    return False
```